Add opkg/NILinuxRT 'reboot required' witnessing feature

10ne1 opened the pull request:
### What does this PR do?

Add  new reboot required witnessed logic to opkg operations on NILRT systems similar to how win_system.py does it, using the same system API as on windows.

### What issues does this PR fix or reference?

Doesn't have an issue assigned.

### New Behavior

A new Linux system API is added for getting/setting reboot witness events which is called by the opkg module. The witness events are stored in volatile memory such that they're cleared upon reboot.

### Tests written?

No

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

Commit:
restartcheck.py: Verify older NILRT kernels

Older NILRT distro versions don't use package management when
installing the kernel bzImage, nor symlinks to files containing the
version, so it's impossible to determine the version without using
ugly hacks like this "strings" one.

The need to maintain backwards compatibility complicates the matter
further because there's no easy way out: We need to wait until support
for older NILRT's is discontinued to avoid having both this hack and
the fixed older NILRT solution at the same time, besides the one
already working on newer distros..

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

Commit:
restartcheck: verify NILRT kernel modules

NI's modules are special in the sense that once they're inserted in
a kernel they can't be removed so a reboot is required when upgrading
or uninstalling them. Some even ask for reboots when installing for
the first time.

NI kernel modules being a given source of reboots, we have to make
restartcheck aware of them by checking a md5sum of modules.dep and
its timestamp (modules.dep is modified/touched by all modules).

The timestamp is checked because on a module upgrade, the modules.dep
contents will be the same (identical md5 hashes).

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

Commit:
opkg: add minion reboot & service restart support

Some packages on NILRT systems, especially kernel modules which are a
poignant exception in the Linux world for this use case, require
rebooting the minion after an opkg install/remove/upgrade/etc.

This adds support for tracking kernel module states to know when a
system reboot is necessary (via restartcheck) and calls a 'witness'
function to set a temporary 'reboot required' state which is cleared
on boot.

If restartcheck output indicates services require restart without
rebooting the system, said services are restarted without calling the
'reboot witness'.

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

Commit:
restartcheck: update function doc

The doc comment is innacurate: in case of error, the returned
dictionary will have the 'result' key set to False and if no
packages are found to need restart, a string is returned.

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

Commit:
system: add reboot API with inital NILRT implementation

This adds a UNIX 'reboot required' tracking mechanism similar to the
one in win_system via two witness get/set functions (same API).

Only NILinuxRT makes use of it for now via the opkg module.

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

skizunov reviewed (approved):
There are a few instances of inconsistent use of single-quotes (') vs doublequotes ("), but I don't know if that qualifies as an issue. It looks fine overall.

cachedout commented on salt/modules/opkg.py:
```diff
@@ -412,12 +453,19 @@ def install(name=None,
                                   'new': new.get(pkgname, '')}})
 
+    rs_result = __salt__['restartcheck.restartcheck'](verbose=False)
+    if isinstance(rs_result, dict) and 'comment' in rs_result:
```
We have this duplicated over on L527. Could we refactor it to a single block that's called?

10ne1 replied:
Done.

cachedout commented on salt/modules/system.py:
```diff
@@ -608,3 +608,54 @@ def get_computer_name():
+
+        salt '*' system.set_reboot_required_witnessed
+    '''
+    if __grains__.get('os_family') == 'NILinuxRT':
```
Hrm. I don't like that this function is essentially a no-op for everything _but_ the `NILinuxRT` OS family but exists in the `system` module which is available and consumed by all types of systems. Is this the best place for this?

skizunov replied:
@cachedout : I think the idea behind this was to copy from the architecture already put in place in `salt/modules/win_system.py`, which is also a "system" module (by virtual name). So the function names and location are not new. I believe this exact same feature could apply to other Linux distributions, I would assume that the author was not motivated to implement and test it on other distributions. Although, by looking at things, it seems that it wouldn't be difficult to do so.

Perhaps the author should add at least one more distribution (eg CentOS) so that your automated tests would cover this new feature?

10ne1 replied:
I can move it somewhere else, no problem, we just need to figure out where is the best place to put them. How about the opkg module? It was written for the NILinuxRT family and it's the only place which calls these witness functions.

terminalmage replied:
I think that opkg.py is probably the correct place to put it. I don't see any code here that actually calls these new functions, and since they only apply to NILinuxRT they shouldn't be in a more general module.

skizunov replied:
@terminalmage : I disagree, because I think these new functions are not private and really meant to be called at the job level. Eg `system.get_reboot_required_witnessed`. How else would the master find out if a reboot is required on a given minion? Because they are required to be called at the job level, putting them into a module with virtual name `pkg` doesn't make that much sense, because all modules with virtual name `pkg` should adhere to a predefined set of APIs, which this falls outside of.

Since there already is precedent for `system.get_reboot_required_witnessed` from the Windows side, it seems like the logical place to put these functions where they are right now. This feature can easily be extended to other Linux distributions, because the main logic that checks if a reboot is required is in `salt/modules/restartcheck.py`, which already supports Debian and Red Hat based systems.

The `pkg` modules for such systems could be modified in a similar way to check if a restart is required after a package operation, and invoke `system.set_reboot_required_witnessed` in such a case. Then, for such systems, a user has the option of checking if a reboot is required after a package operation (eg if a new kernel is installed). It would actually be a pretty useful feature for general use.

10ne1 replied:
I'm ok either way, I think we just need a consensus so I know what to do further.

terminalmage replied:
@skizunov Thanks, I overlooked your earlier comment regarding the `win_system` module. Sorry about that! I agree, system.py is a good place for it.

cachedout commented on salt/modules/system.py:
```diff
@@ -608,3 +608,54 @@ def get_computer_name():
+
+        salt '*' system.get_reboot_required_witnessed
+    '''
+    if __grains__.get('os_family') == 'NILinuxRT':
```
Same concern as above here.

cachedout commented:
@terminalmage Re-review requested.

terminalmage reviewed (approved):
@cachedout sorry, my comment [here](https://github.com/saltstack/salt/pull/46774#discussion_r179514086) was meant to signify that my concerns were satisfied, but I neglected to submit an "Approved" review. Sorry for holding things up.

terminalmage commented on salt/modules/system.py:
```diff
@@ -608,3 +608,54 @@ def get_computer_name():
+    if __grains__.get('os_family') == 'NILinuxRT':
+        if os.path.exists(NILRT_REBOOT_WITNESS_PATH):
+            return True
+    return False
```
Actually, I like the idea of such decorators even if we don't use them here, so I am looking into writing them now.

isbm replied:
@terminalmage :+1: for the decorator. I would still map it separately into its own namespace. Because it is "special" stuff. Then in your state you can better segregate these calls, like:

```
salt * somemodule.special.do_something
```
Then you will also construct your states/commands differently, avoiding hitting systems that aren't in this space.

isbm replied:
@skizunov are we on the same page? Sounds like you're talking about an interface. And it is yet to be implemented. What I am talking is a principal of an actual implementation of _any_ module that on some OSes even do not have to exist. That said, if other systems needs to implement this, then exception raise "not yet implemented" or "cannot be at all supported" is needed here.

But right now it is returning one of an identical return data of the actual implementation. And so this is a bug and confusion.

skizunov replied:
@isbm : This interface has already been implemented for Windows systems (in `win_system.py` which also implements the `system` virtual interface). But I understand your point about the behavior on other Linux distributions that may not necessarily implement the interface. So perhaps there are a few options:
1) Use the solution that @terminalmage suggests (requires infrastructure not yet present)
2) Just implement for the other Linux distributions (should be very easy, just need to confirm the path to volatile storage is valid for each Linux distribution)
3) On other Linux distributions that don't (yet) support this, throw an exception instead of returning a value.

isbm replied:
@skizunov I'd prefer @terminalmage 's suggestion as it is case-agnostic.

terminalmage replied:
Implementing function namespaces is not going to happen right now. It may in a future release, but it's extremely unlikely that it will happen in time for Fluorine. So, we should adjust expectations here accordingly.

Late last week I added the needed support to the `depends` decorator to enable the running of a shell command (https://github.com/saltstack/salt/pull/47040). In the process of doing so, I discovered that this decorator _already_ supports interpreting an expression that evaluates to a boolean (with `True` meaning that the function should be loaded and `False` meaning that it should not).

There are two approaches we could use:

**OPTION 1 (depends on https://github.com/saltstack/salt/pull/47040, would have to wait for that PR to be merged and then rebase to gain access to this functonality)**

```python
from salt.utils.decorators import depends

@depends('some shell command', retcode=0)
def set_reboot_required_witnessed():
    '''
```

**OPTION 2**

```python
from salt.utils.decorators import depends

@depends(some python expression)
def set_reboot_required_witnessed():
    '''
```

Option 2 could be done right now without waiting for my decorator modifications to be merged, but note that whatever is passed to the depends decorator must evaluate as boolean. So it has to be an expression or function that returns a bool.

The question is, I guess, what would be the best way to confirm that the minion is NILinuxRT? Keep in mind that we won't have access to the grains.

10ne1 replied:
I've rebased on top of master and have the code from #47040 

Can I run any shell commands in the depends decorator? I can check if we're on NILRT using 'lsb-release -i | grep -q nilrt ' if yes.

cachedout commented:
I think we're still waiting on some feedback for some of the concerns raised by @isbm here.

isbm commented:
@10ne1 any updates here?

10ne1 commented:
Sorry for the delay, I've been overwhelmed with other work + conference travel, I'll get back to this ASAP (I hope to update the PR tomorrow).

10ne1 commented on salt/modules/opkg.py:
```diff
@@ -55,12 +56,62 @@
+        if not (os.path.exists(os.path.join(NILRT_MODULE_STATE_PATH, 'modules.dep.timestamp')) and
+                os.path.exists(os.path.join(NILRT_MODULE_STATE_PATH, 'modules.dep.md5sum'))):
+            try:
+                os.makedirs(NILRT_MODULE_STATE_PATH)
```
Done.

10ne1 commented on salt/modules/opkg.py:
```diff
@@ -55,12 +56,62 @@
+            except OSError as exc:
+                if exc.errno != errno.EEXIST:
+                    msg = 'Error creating {0} (-{1}): {2}'.format(NILRT_MODULE_STATE_PATH, exc.errno, exc.strerror)
+                    return (False, msg)
```
Done.

10ne1 commented on salt/modules/restartcheck.py:
```diff
@@ -298,6 +298,13 @@ def _kernel_versions_redhat():
+    '''
+    If this is an older version of NILinuxRT, return True. Otherwise, return False.
+    '''
+    return bool(os.path.exists('/usr/local/natinst/bin/nisafemodeversion'))
```
Done.

10ne1 commented on salt/modules/restartcheck.py:
```diff
@@ -307,14 +314,61 @@ def _kernel_versions_nilrt():
-    kernel = os.path.basename(kernel)
-    kernel = kernel.strip('bzImage-')
+
+    if __grains__.get('os_family') == 'NILinuxRT' and _is_older_nilrt():
```
Done.

10ne1 commented on salt/modules/restartcheck.py:
```diff
@@ -307,14 +314,61 @@ def _kernel_versions_nilrt():
+        # backwards compatible so it'll just get more complicated.
+        kpath = '/boot/runmode/bzImage'
+        kvregex = r'[0-9]+\.[0-9]+\.[0-9]+-rt'
+        kernel = __salt__['cmd.shell']('strings {0} | awk \'$1 ~ /{1}/ {{print $1}}\''
```
Done (ported the awk logic to python re)

10ne1 commented on salt/modules/restartcheck.py:
```diff
@@ -307,14 +314,61 @@ def _kernel_versions_nilrt():
+    else:
+        kernel = os.readlink('/boot/bzImage')
+        kernel = os.path.basename(kernel)
+        kernel = kernel.strip('bzImage-')
```
Done.

10ne1 commented on salt/modules/restartcheck.py:
```diff
@@ -307,14 +314,61 @@ def _kernel_versions_nilrt():
+    if retcode != 0:
+        return True
+
+    return False
```
Done.

10ne1 commented on salt/modules/restartcheck.py:
```diff
@@ -359,8 +413,14 @@ def restartcheck(ignorelist=None, blacklist=None, excludepid=None, verbose=True)
         if kernel in kernel_current:
-            kernel_restart = False
-            break
+            if __grains__.get('os_family') == 'NILinuxRT':
```
Done.

isbm commented:
@10ne1 do not forget to push. :wink:

10ne1 commented:
@isbm Sure, I'll push, I just need to do a little more testing (got preempted again but I'm back on it :smiley: )

isbm commented on salt/modules/restartcheck.py:
```diff
@@ -298,6 +300,13 @@ def _kernel_versions_redhat():
+    '''
+    If this is an older version of NILinuxRT, return True. Otherwise, return False.
+    '''
+    return os.path.exists('/usr/local/natinst/bin/nisafemodeversion')
```
This is strange. You're judging _version_ by just a mere place of the binary installation. Why you're so sure that the `nisafemodeversion` isn't up to the latest _version_ but just is a symlink for backward compatibility? If you want to look for version, then look for the version by actually calling it or refer to some data that would actually fetch the real version and compare it.

10ne1 replied:
We have two big incompatible versions (usually we call them older and newer NILRT) of the distro supported side by side and there is no way to distinguish them from the rootfs (no versioning info to make that distinction in any package). I could implement a machanism to do that but we'd still have to support already released images...

That nisafemodeversion file is not present t all in newer NILRT, as we have a unified bootflow with no "safe mode", so even though it looks ugly that's the most reliable way I could come up to distinguish older vs newer :sweat_smile:

isbm replied:
@10ne1 I hear you. It is still would be nicer if your distro would start using `/etc/os-release` like everyone else are doing.

isbm commented on salt/modules/system.py:
```diff
@@ -608,3 +609,54 @@ def get_computer_name():
+                                      .format(dir_path, ex.errno, ex.strerror))
+
+        __salt__['cmd.run']('touch {0}'.format(NILRT_REBOOT_WITNESS_PATH))
+        return True
```
And if that fails for whatever reason with error code non zero, you're still saying "all perfect". I would not return it here, but at the end, explicitly that the exit code is 0 and then bool it:

```python
errcode = -1
...
if not os.path.exists(dir_path):
    ...
    errcode = ....

return errcode == 0
```

10ne1 replied:
Done.

isbm commented on salt/modules/system.py:
```diff
@@ -608,3 +609,54 @@ def get_computer_name():
+
+NILRT_REBOOT_WITNESS_PATH = '/var/volatile/tmp/salt/reboot_witnessed'
+
+NILRT_FAMILY_VERIFY_CMD = 'sh -c \'lsb_release -d | grep -qi "NI Linux Real-Time"\''
```
It has to be better than this...
```
$ lsb_release -d
-bash: lsb_release: command not found
```

10ne1 replied:
Done, I'm using the os family grain now.

isbm reviewed (changes requested):
Sorry, I found some more concerns. Could you look at them?

Commit:
restartcheck: fix initscripts detection

The fallback if's are reversed :)

__salt__['service.available'] currently returns only initscripts so
don't report them as systemd services by default (check first to see
if it's an initscript and if not add them to systemd because in the
future __salt__['service.available'] may return systemd services).

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

10ne1 commented:
Rebased (resolved 1 conflict with develop branch) and added 1 more fix.
